File: scala-server/test/com/bolour/util/scala/common/CompactStringSetSpec.scala
/*
 * Copyright 2017-2018 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */

package com.bolour.util.scala.common

import scala.io.{Source, BufferedSource}
import org.scalatest.{FlatSpec, Matchers}

class CompactStringSetSpec extends FlatSpec with Matchers {

  val runtime = Runtime.getRuntime

  def printMemory = {
    val memory = runtime.totalMemory()
    val freeMemory = runtime.freeMemory()
    val usedMemory = memory - freeMemory

    println(s"used memory: ${usedMemory}, free memory: ${freeMemory}")
  }

  "compact string set" should "store and retrieve strings" in {
    val set = new CompactStringSet()
    set.insert("one")
    set.insert("two")
    set.contains("one") shouldBe true
    set.contains("two") shouldBe true
    set.contains("three") shouldBe false

    set ++ List("four", "five").toIterator
    set.contains("four") shouldBe true
    set.contains("five") shouldBe true
    set.contains("one") shouldBe true

  }

  // Memory and timing test. Time-consuming.
//  "compact string set" should "have little memory overhead" in {
//    val begin = System.currentTimeMillis()
//    val source = Source.fromFile("dict/en-masked-words.txt")
//    val maskedWords = source.getLines
//    printMemory
//    val set = new CompactStringSet()
//    set ++ maskedWords
//    val end = System.currentTimeMillis()
//    val time = (end - begin)/1000
//    println(s"time: ${time} seconds")
//    printMemory
//    println(s"number of buckets: ${set.buckets.size}")
//
//  }

}

File: haskell-server/test/Bolour/Util/CompactStringSetSpec.hs
--
-- Copyright 2017-2018 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}

module Bolour.Util.CompactStringSetSpec where

import Test.Hspec

import qualified Data.Map.Strict as Map
import Data.Time (UTCTime, getCurrentTime)
import Data.Time.Clock (diffUTCTime)
import Control.Concurrent (threadDelay)

import qualified Bolour.Util.CompactStringSet as Set

spec :: Spec
spec = do
  describe "compact string set" $ do
    let set = Set.mkCompactStringSet
    it "should store and retrieve strings" $ do
      let set' = Set.insert set "one"
      Set.contains set' "one" `shouldBe` True
      let set'' = Set.insertAll set' ["two", "three"]
      Set.contains set'' "two" `shouldBe` True
      Set.contains set'' "three" `shouldBe` True
      Set.contains set'' "four" `shouldBe` False

      -- Insertion of bulk data uses excessive memory. Not sure why. Give up for now.
--   describe "timing of insertions into compact string set" $
--     -- let set = Set.mkCompactStringSet
--     it "insert large lists in reasonable time" $ do
--       threadDelay (1 * 1000000)
--       begin <- getCurrentTime
--
--       contents <- readFile "dict/en-masked-words.txt"
--       let strings = lines contents
--       -- let s = strings !! 10000000
--       -- print s
--       let set @ Set.CompactStringSet { buckets } = Set.mkFromList strings
--       -- print $ "number of buckets: " ++ show (Map.size buckets)
--       print $ "contains washy: " ++ show (Set.contains set "washy")
--
--       threadDelay (1 * 5000000)
--
--       end <- getCurrentTime
--       let diffTime = diffUTCTime end begin
--       let seconds = toInteger $ floor diffTime
--       print $ "time " ++ show seconds ++ " seconds"
--       True `shouldBe` True





